go-pkg/ds/permission.go

```diff
@@ -22,6 +28,55 @@ func (u *User) addKey(blob *protods.DataBlob, newDataOwner *User) *protods.DataB
 // 1. Generate an AES-key
 // 2. Add an entry to "Secrets"
 // 3. Encrypt data with AES key
-func (u *User) createDataBlob(data []byte) *protods.DataBlob {
-	panic("not implemented")
+func (u *User) createDataBlob(data []byte) (*protods.DataBlob, error) {
+	//HiddenDataKey creation
+	dataKey := encrypt.GenAes128Key()
+
+	ephemeralPrivKey, err := ethereum.GenerateKey()
+	if err != nil {
+		return nil, errors.New("creation of ephemeral private key failed")
+	}
+	sharedSecretX, sharedSecretY := ephemeralPrivKey.PublicKey.Add(
+		u.privkey.X,
+		u.privkey.Y,
+		ephemeralPrivKey.PublicKey.X,
+		ephemeralPrivKey.PublicKey.Y,
+	)
+
+	sharedSecret := elliptic.Marshal(ethereum.S256(), sharedSecretX, sharedSecretY)
+	sharedSecretHash := sha256.Sum256(sharedSecret)
+
+	iv := encrypt.GenCBCIv()
+
+	encryptedHiddenKey, err := encrypt.Encrypt(sharedSecretHash[:], iv, dataKey)
+	if err != nil {
+		return nil, errors.New("signing failed")
+	}
+
+	hiddenDataKey := protods.SharedKey{
+		Pubkey:          ethereum.CompressPubkey(&u.privkey.PublicKey),
+		EphemeralPubkey: ethereum.CompressPubkey(&ephemeralPrivKey.PublicKey),
+		EncryptedKey:    encryptedHiddenKey,
+		Iv:              iv,
+	}
+
+	//DataBlob creation
+	dataLen := len(data)
+
+	encryptedData, err := encrypt.Encrypt(dataKey, iv, data)
```

Comment:
Do not re-use `iv`. They are two separate encryptions. They should get their own iv.